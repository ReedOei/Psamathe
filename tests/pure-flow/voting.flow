#nopreprocess

type Voter is unique immutable asset nat

type ProposalName is unique immutable asset nat

type Election is asset {
    eligibleVoters : any list one Voter,
    proposals : any map one ProposalName => any list one Voter
}

transformer createProposal(this : one Election, proposal : any nat) -> success : one bool {
    new ProposalName(proposal) --> var name : ProposalName
    [ one Voter ; ] --> var voters : list one Voter
    [ one record(key) { key : one ProposalName, value : any list one Voter } ;
        record(key) {
            key : one ProposalName |-> name,
            value : any list one Voter |-> voters
        }
    ] --> this.proposals
}

transformer giveRightToVote(this : one Election, voter : any nat) -> success : one bool {
    new Voter(voter) --> this.eligibleVoters
    true --> success
}

transformer vote(this : one Election, voter : any nat, proposal : any nat) -> success : one bool {
    this.eligibleVoters --[ voter ]-> this.proposals[proposal].value
    true --> success
}

;

[ one Voter ; ] --> var eligibleVoters : list one Voter
[ one record(key) { key : one ProposalName, value : any list one Voter} ; ] --> var proposals : map one ProposalName => any list one Voter
new Election(eligibleVoters, proposals) --> var election : Election

createProposal(election, 0) --> var success : bool
/* giveRightToVote(election, 0) --> success */
giveRightToVote(election, 1) --> success
vote(election, 1, 0) --> success

