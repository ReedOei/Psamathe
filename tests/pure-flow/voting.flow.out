<generatedTop>
  <common>
    <k>
      .
    </k>
    <typeEnv>
      .Map
    </typeEnv>
    <types>
      Election |-> typeDef ( SetItem ( asset ) , record ( .Vars ) { eligibleVoters : any table ( .Vars ) one Voter , proposals : any table ( key , .Vars ) one record ( key , .Vars ) { key : one ProposalName , value : any table ( .Vars ) one Voter , .VarDefs } , .VarDefs } )
      ProposalName |-> typeDef ( SetItem ( asset )
      SetItem ( immutable )
      SetItem ( unique ) , nat )
      Voter |-> typeDef ( SetItem ( asset )
      SetItem ( immutable )
      SetItem ( unique ) , nat )
    </types>
    <functions>
      createProposal |-> transformer ( success : one bool , this : one Election , proposal : any nat , .VarDefs , proposal --> new ProposalName ( .Vars ) --> var name : ProposalName  [ one Voter ; .Vars ] --> var voters : table ( .Vars ) one Voter  [ one record ( key , .Vars ) { key : one ProposalName , value : any table ( .Vars ) one Voter , .VarDefs } ; record ( key , .Vars ) { key : one ProposalName |-> name , value : any table ( .Vars ) one Voter |-> voters , .RecordMembers } , .Vars ] --> this . proposals  .Stmts )
      giveRightToVote |-> transformer ( success : one bool , this : one Election , voter : any nat , .VarDefs , voter --> new Voter ( .Vars ) --> this . eligibleVoters  true --> success  .Stmts )
      vote |-> transformer ( success : one bool , this : one Election , voter : any nat , proposal : any nat , .VarDefs , this . eligibleVoters [ voter ] --> this . proposals [ proposal ] . value  true --> success  .Stmts )
    </functions>
  </common>
  <lookup>
    election |-> selected ( loc ( 2 ) , loc ( 2 ) )
    eligibleVoters |-> selected ( loc ( 0 ) , loc ( 0 ) )
    proposals |-> selected ( loc ( 1 ) , loc ( 1 ) )
    success |-> selected ( loc ( 7 ) , loc ( 7 ) )
  </lookup>
  <storage>
    loc ( 0 ) |-> resource ( table ( .Vars ) one Voter , .List )
    loc ( 1 ) |-> resource ( table ( key , .Vars ) one record ( key , .Vars ) { key : one ProposalName , value : any table ( .Vars ) one Voter , .VarDefs } , .List )
    loc ( 2 ) |-> resource ( Election , ListItem ( eligibleVoters |-> loc ( 5 )
    proposals |-> loc ( 6 ) ) )
    loc ( 3 ) |-> resource ( Election , .List )
    loc ( 4 ) |-> resource ( record ( .Vars ) { eligibleVoters : any table ( .Vars ) one Voter , proposals : any table ( key , .Vars ) one record ( key , .Vars ) { key : one ProposalName , value : any table ( .Vars ) one Voter , .VarDefs } , .VarDefs } , .List )
    loc ( 5 ) |-> resource ( table ( .Vars ) one Voter , ListItem ( loc ( 20 ) ) )
    loc ( 6 ) |-> resource ( table ( key , .Vars ) one record ( key , .Vars ) { key : one ProposalName , value : any table ( .Vars ) one Voter , .VarDefs } , ListItem ( loc ( 16 ) ) )
    loc ( 7 ) |-> resource ( bool , ListItem ( true ) )
    loc ( 8 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 9 ) |-> resource ( bool , .List )
    loc ( 10 ) |-> resource ( ProposalName , .List )
    loc ( 11 ) |-> resource ( ProposalName , .List )
    loc ( 12 ) |-> resource ( table ( .Vars ) one Voter , .List )
    loc ( 13 ) |-> resource ( record ( key , .Vars ) { key : one ProposalName , value : any table ( .Vars ) one Voter , .VarDefs } , .List )
    loc ( 14 ) |-> resource ( ProposalName , ListItem ( 0 ) )
    loc ( 15 ) |-> resource ( table ( .Vars ) one Voter , ListItem ( loc ( 20 ) ) )
    loc ( 16 ) |-> resource ( record ( key , .Vars ) { key : one ProposalName , value : any table ( .Vars ) one Voter , .VarDefs } , ListItem ( key |-> loc ( 14 )
    value |-> loc ( 15 ) ) )
    loc ( 17 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 18 ) |-> resource ( bool , .List )
    loc ( 19 ) |-> resource ( Voter , .List )
    loc ( 20 ) |-> resource ( Voter , ListItem ( 1 ) )
    loc ( 21 ) |-> resource ( bool , .List )
    loc ( 22 ) |-> resource ( nat , ListItem ( 0 ) )
    loc ( 23 ) |-> resource ( bool , .List )
    loc ( 24 ) |-> resource ( nat , ListItem ( 1 ) )
    loc ( 25 ) |-> resource ( bool , .List )
  </storage>
  <catchStack>
    .List
  </catchStack>
  <createdVals>
    ProposalName |-> createdVals ( ListItem ( resource ( nat , ListItem ( 0 ) ) ) )
    Voter |-> createdVals ( ListItem ( resource ( nat , ListItem ( 1 ) ) ) )
  </createdVals>
</generatedTop>
